base/compiler/typelimits.jl

```diff
@@ -349,7 +349,7 @@ function tmerge(@nospecialize(typea), @nospecialize(typeb))
     # if we didn't start with any unions, then always OK to form one now
     if !(typea isa Union || typeb isa Union)
         # except if we might have switched Union and Tuple below, or would do so
-        if (isconcretetype(typea) && isconcretetype(typeb)) || !(typea <: Tuple && typeb <: Tuple)
+        if (isconcretetype(typea) && isconcretetype(typeb)) || !(typea <: Tuple || typeb <: Tuple)
```

Comment:
This was the minimal change to make the recursive test case (`f27316` below) work. It's rather ad-hoc though. Could we just simplify this to the `inconcretetype` part?

Reply:
Yes (the second test was wrong, it should have been `typeintersect`)

Reply:
No, `isconcretetype(typea) && isconcretetype(typeb)` is a little too strict, it makes these tests fail:
https://github.com/JuliaLang/julia/blob/83ce7ba119ae49f968a74f64e423e0ff5c2fbc71/test/compiler/compiler.jl#L334
https://github.com/JuliaLang/julia/blob/83ce7ba119ae49f968a74f64e423e0ff5c2fbc71/test/compiler/compiler.jl#L383
We probably want to allow `isType(...)`, too.